fix: bottom navigation bar colors

talamaska opened the pull request:
add: _effectiveIconTheme function
add: logic for setting the effective selected/unselected icon theme 
change: use the new effective selected/unselected Icon theme when generating _BottomNavigationTile
change: create separate tweens for the icon and for the label
add: tests to capture the issues with the label style color, matching the docs
fix: BottomNavigationBarItem label's color is not overridden neither by selectedLabelStyle nor unselectedLabelStyle
add: legacy flag for preserving the old behavior

#64358
#107925
#27753

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

darrenaustin commented on packages/flutter/lib/src/material/bottom_navigation_bar.dart:
```diff
@@ -381,6 +385,10 @@ class BottomNavigationBar extends StatefulWidget {
   ///    orientation.
   final BottomNavigationBarLandscapeLayout? landscapeLayout;
 
+  /// This flag is controlling how [BottomNavigationBar] is going to use
+  /// the colors provided by the properties and the theme.
```
We should have a more complete description of exactly what this flag does.

darrenaustin replied:
Is this needed because it breaks existing tests if it is not true by default?

talamaska replied:
Yes, it is a change that would break your golden test. You could refer to the closed PR for our conversation on that matter. The flag was suggested by @Hixie in order to avid that and still have the changes implemented. The goldens break as soon I separate the color Tweens. This was a suggestion from you in the previous closed PR. It is correct a suggestion.

darrenaustin commented on packages/flutter/lib/src/material/bottom_navigation_bar.dart:
```diff
@@ -910,24 +920,24 @@ class _BottomNavigationBarState extends State<BottomNavigationBar> with TickerPr
+  // Otherwise, the [IconThemeData]'s color should be selectedItemColor
+  // or unselectedItemColor.
+  static IconThemeData _effectiveIconTheme(IconThemeData? iconTheme, Color? itemColor) {
+    final IconThemeData effectiveIconTheme = iconTheme ?? const IconThemeData();
+    // Prefer the iconTheme over itemColor and textStyle color if present.
+    // If iconTheme is not set, prefer textStyle color over itemColor
+    return iconTheme == null ? effectiveIconTheme.copyWith(color: itemColor) : effectiveIconTheme;
```
Couldn't this just be:

```suggestion
    return iconTheme ?? IconThemeData(color: itemColor);
```

talamaska replied:
I'll make sure this doesn't break the unit tests, cause I remember we were going through that already.

darrenaustin commented on packages/flutter/lib/src/material/bottom_navigation_bar.dart:
```diff
@@ -154,6 +154,9 @@ class BottomNavigationBar extends StatefulWidget {
   /// [selectedIconTheme] and [unselectedIconTheme], and both
   /// [IconThemeData.color] and [IconThemeData.size] must be set.
   ///
+  /// [IconThemeData]s take precedence over [iconSize], [selectedItemColor]
+  /// and [unselectedItemColor] for icons.
+  ///
```
```suggestion
  /// [selectedIconTheme] values will be used instead of [iconSize] and [selectedItemColor] for selected icons.
  /// [unselectedIconTheme] values will be used instead of [iconSize] and [unselectedItemColor] for unselected icons.
  ///
```

darrenaustin replied:
And isn't this only true if `useLegacyColorScheme` is false?

talamaska replied:
correct, I should mention this in the docs right?

darrenaustin reviewed (changes requested):
@talamaska thanks for the contribution. Sorry for the lag on the review.

This mostly looks good. I am not a fan of the `useLegacyColorScheme` type flags, but I assume this is to work around a breaking change?

talamaska commented:
@darrenaustin implemented the suggestions. Something is wrong with the ci. All items are showing failed authentication. I'm not sure what I can do about it.

talamaska committed:
fix: bottom navigation bar colors

add: _effectiveIconTheme function
add: logic for setting the effective selected/unselected icon theme
change: color tweens to use the effective label style color if not null
change: use the new effective selected/unselected Icon theme when generating _BottomNavigationTile
add: tests to capture the issues with the label style color, matching the docs
fix: BottomNavigationBarItem title's color is not overriden neither by selectedLabelStyle nor unselectedLabelStyle
remove: trailing space


implement PR suggestions

add: more documentation for the useLegacyColorScheme flag
change: simplify effectiveIconTheme

talamaska committed:
Merge remote-tracking branch 'origin/fix/64358-bottom-navigation-bar-colors' into fix/64358-bottom-navigation-bar-colors

talamaska commented:
@darrenaustin @Hixie Am I going to wait for another 6 months for someone to finalize the review and merge this PR. The contributing experience so far is not the greatest one. Do I need to copy the source of this widget and release it in pub.dev, so that I can finally use a non-broken and not wrapped with custom logic. I'm sorry but I'm running out of patience soon and this is what I'm going to do, even if that wasn't my intend. All the devrel, flutter community representative proclaim that everybody can help making Flutter better. But this is not my experience. The initial work for this fix was done in January.

darrenaustin reviewed (approved):
LGTM.

darrenaustin commented:
> @darrenaustin @Hixie Am I going to wait for another 6 months for someone to finalize the review and merge this PR. The contributing experience so far is not the greatest one. 

@talamaska my sincerest apologies. This is entirely my fault for not letting this languish so long.  I should have responded in a more timely manner. I will strive to do better in the future. We do appreciate your contribution. Thx.

talamaska commented:
> @talamaska my sincerest apologies. This is entirely my fault for letting this languish so long. I should have responded in a more timely manner. I will strive to do better in the future. We do appreciate your contribution. Thx.

@darrenaustin thank you for responding. I apologize for my harsh comment. In my defense I really tried to be patient to this point

talamaska commented:
@darrenaustin when can I expect this merged?

darrenaustin commented:
> @darrenaustin thank you for responding. I apologize for my harsh comment. In my defense I really tried to be patient to this point

Not at all, it was completely justified frustration. Again, my apologies.

> @darrenaustin when can I expect this merged?

Done. Not sure why it didn't automerge after the review, I thought the label was set on it.

talamaska commented:
@darrenaustin any idea when this can end up in stable? I don't see it in Beta as well. It is still just in master

rscircus commented:
Hm, just ran into this, too. It's a rather old bug. Interesting.

talamaska commented:
> Hm, just ran into this, too. It's a rather old bug. Interesting.

It was released in stable. Could you please try setting useLegacyColorScheme to false. The issue should be resolved.

rscircus commented:
Moved to SnakeNavigationBar meanwhile...

sejun2 commented:
https://api.flutter.dev/flutter/material/BottomNavigationBar/useLegacyColorScheme.html
This explains that.
